src/lenskit/data/matrix.py

```diff
@@ -542,3 +543,46 @@ def _check_index_type(index_type: pa.DataType, dimension: int | None) -> int:
 pa.register_extension_type(SparseIndexType(0))  # type: ignore
 pa.register_extension_type(SparseIndexListType(0))  # type: ignore
 pa.register_extension_type(SparseRowType(0))  # type: ignore
+
+
+def normalize_matrix(
+    matrix: csr_array | NDArray[np.floating[Any]],
+    normalize: Literal["unit", "distribution"] | None,
+) -> csr_array | NDArray[np.floating[Any]]:
+    """
+    Normalize rows of a matrix.
+
+    Args:
+        matrix: Sparse or dense matrix to normalize
+        normalize: Normalization mode ("unit" for L2, "distribution" for L1)
+
+    Returns:
+        Normalized matrix
+    """
+    if normalize is None:
+        return matrix
+
+    is_sparse = isinstance(matrix, csr_array)
+
+    if normalize == "unit":
+        if is_sparse:
+            stats = np.sqrt(np.asarray(matrix.multiply(matrix).sum(axis=1)).ravel())
+        else:
+            stats = np.linalg.norm(matrix, axis=1)
+    else:
+        if (matrix.data if is_sparse else matrix).min() < 0:
+            raise ValueError("Cannot normalize to distribution: negative values present")
+        if is_sparse:
+            stats = np.asarray(matrix.sum(axis=1)).ravel()
+        else:
+            stats = matrix.sum(axis=1)
+
+    stats[stats == 0] = 1.0
+
+    if is_sparse:
+        matrix = matrix.copy()
+        matrix.data = matrix.data.astype(np.float32)
+        matrix.data /= np.repeat(stats, np.diff(matrix.indptr))
```

Comment:
can this be done with `matrix / stats[:, None]` like the dense version? If not, does the `.divide` method work?